### ISSUE TITLE
[java] AvoidArrayLoops regression: from false negative to false positive with final variables

### ISSUE THREAD

**Issue by Androbin** [pmd/pmd#4183]
**Affects PMD Version:** 6.51.0

**Rule:** [AvoidArrayLoops](https://pmd.github.io/latest/pmd_rules_java_performance.html#avoidarrayloops)

**Description:**
While fixing #3847, a regression was introduced in #4159 @adangel, now instead of a false negative we have a false positive.
A final variable may only be ignored if it's computed before the loop (or otherwise constant over the loop).

**Code Sample demonstrating the issue:**

```java
public class Test {
  public static void sample() {
    final int[] a = new int[10];
    final int[] b = new int[10];
    for (int i=0; i<10; i++) {
      final int c = i;
      b[i] = a[i+c];
    }
  }
}
```

**Expected outcome:**

PMD reports a violation at line 5, but that's wrong. That's a false positive.

**Running PMD through:** *CLI*

<details>
  <summary>True positive code samples</summary>
  
```java
public class Test {
  public static void sample() {
    final int[] a = new int[10];
    final int[] b = new int[10];
    final int c = 6;
    for (int i=0; i<10; i++) {
      b[i] = a[i+c];
    }
  }
}
```
  
```java
public class Test {
  public static void sample() {
    final int[] a = new int[10];
    final int[] b = new int[10];
    for (int i=0; i<10; i++) {
      final int c = 6;
      b[i] = a[i+c];
    }
  }
}
```
  
</details>
<details>
  <summary>True negative code samples</summary>
  
```java
public class Test {
  public static void sample() {
    final int[] a = new int[10];
    final int[] b = new int[10];
    for (int i=0; i<10; i++) {
      int c = i;
      b[i] = a[i+c];
    }
  }
}
```
  
</details>
